community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreStatement.java

```diff
@@ -0,0 +1,260 @@
+/*
+ * Copyright (c) 2002-2015 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.api.store;
+
+import org.neo4j.kernel.impl.store.NeoStore;
+import org.neo4j.kernel.impl.store.NodeStore;
+import org.neo4j.kernel.impl.store.PropertyStore;
+import org.neo4j.kernel.impl.store.RelationshipGroupStore;
+import org.neo4j.kernel.impl.store.RelationshipStore;
+import org.neo4j.kernel.impl.store.record.NodeRecord;
+import org.neo4j.kernel.impl.store.record.RelationshipGroupRecord;
+import org.neo4j.kernel.impl.store.record.RelationshipRecord;
+
+/**
+ * Statement for store layer. This allows for acquisition of cursors on the store data.
+ * <p/>
+ * The cursors call the release methods, so there is no need for manual release, only
+ * closing those cursor.
+ * <p/>
+ * {@link NeoStore} caches one of these per thread, so that they can be reused between statements/transactions.
+ */
+public class StoreStatement
+        implements AutoCloseable
+{
+    private StoreSingleNodeCursor singleNodeCursor;
+    private StoreIteratorNodeCursor iteratorNodeCursor;
+    private StorePropertyCursor propertyCursor;
+    private StoreLabelCursor labelCursor;
+    private StoreSingleRelationshipCursor singleRelationshipCursor;
+    private StoreNodeRelationshipCursor nodeRelationshipCursor;
+    private StoreIteratorRelationshipCursor iteratorRelationshipCursor;
+    private NeoStore neoStore;
+
+    public StoreStatement( NeoStore neoStore )
+    {
+        this.neoStore = neoStore;
+    }
+
+    public StoreSingleNodeCursor acquireSingleNodeCursor()
+    {
+        if ( singleNodeCursor == null )
+        {
+            NodeStore nodeStore = neoStore.getNodeStore();
+            if ( nodeStore == null )
+            {
+                throw new IllegalStateException( "Database has been shutdown" );
+            }
+
+            singleNodeCursor = new StoreSingleNodeCursor( new NodeRecord( -1 ), nodeStore, this );
+        }
+
+        try
+        {
+            return singleNodeCursor;
+        }
+        finally
+        {
+            singleNodeCursor = null;
+        }
+    }
+
+    void releaseSingleNodeCursor( StoreSingleNodeCursor nodeCursor )
+    {
+        this.singleNodeCursor = nodeCursor;
+    }
```

Comment:
It feels that some abstraction is missing here, there is quite of code duplication